### ISSUE TITLE
Rule 941100 false positive with WordPress

### ISSUE THREAD

**Issue by azurit** [SpiderLabs/owasp-modsecurity-crs#1511]
## Type of Issue
false positive

## Description

WordPress version 5.2.2 with crs_exclusions_wordpress enabled. Rule is triggered multiple times for argument 'element'. I sent the full log to folini@netnea.com.

## Your Environment

* CRS version (e.g. v3.0.2): 3.1.1
* ModSecurity version (e.g. 2.9.2): 2.9.3
* Web Server and version (e.g. apache 2.4.27): 2.4.25
* Operating System and version: Debian Stretch

## Confirmation

[x] I have removed any personal data (email addresses, IP addresses,
    passwords, domain names) from any logs posted.

**Comment by azurit** [comment]
Found out that external theme is causing this (which is using custom visual editor). Guessing third party plugins/themes/whatever aren't supported? :)

**Comment by dune73** [comment]
OK. I took a look. 9 of the 11 alerts in your log file (thanks for sending) are against 941100. This is handled by libinjection, a 3rd party library that ModSecurity uses. If there are false positives we can do very little outside of disabling said rule / library for certain elements.

Here they all target parameters of the form `element[1][0][]` and the like. So one would have to exclude them all for the path `/wp-admin/admin-ajax`. Can you confirm that this is a current version of wordpress?

Furthermore, two hits on 941130 on the same parameter(s).

**Comment by azurit** [comment]
It was a WordPress 5.2.2 with 3th party theme called Mioweb (don't know the version, if it has any). That theme has it's own visual editor included and that was triggering the mentioned rules. I already wrote an exclusion rule for this one (not sure about the 'element' array indices, is it possible to write it better? for example using a regexp):

```
SecRule REQUEST_FILENAME "@endsWith /wp-admin/admin-ajax.php" \
    "id:9990101,\
    phase:2,\
    pass,\
    t:none,\
    nolog,\
    chain"
    SecRule ARGS:action "@streq save_page" \
        "t:none,\
        chain"
        SecRule &ARGS:action "@eq 1" \
            "t:none,\
            ctl:ruleRemoveTargetById=941100;ARGS:element[0][0][],\
            ctl:ruleRemoveTargetById=941100;ARGS:element[0][1][],\
            ctl:ruleRemoveTargetById=941100;ARGS:element[1][0][],\
            ctl:ruleRemoveTargetById=941100;ARGS:element[2][0][],\
            ctl:ruleRemoveTargetById=941100;ARGS:element[3][0][],\
            ctl:ruleRemoveTargetById=941130;ARGS:element[0][1][],\
            ctl:ruleRemoveTargetById=941130;ARGS:element[1][0][],\
            ctl:ruleRemoveTargetById=941100;ARGS:layer,\
            ctl:ruleRemoveTargetById=941130;ARGS:layer"

```

**Comment by dune73** [comment]
That's a well written rule exclusion. No, unfortunately, you have to address them directly like this. I hope the numbers don't change... Other than that you can only disable the rule completely in this situation.

**Comment by albgen** [comment]
Hello,

this rule is blocking me from changing a page in wordpress. I have just installed yesterday the latest CRS and also the latest wordpress.

Any idea?

**Comment by azurit** [comment]
@albgen Use my exclusion rule above.

**Comment by albgen** [comment]
> @albgen Use my exclusion rule above.

hi, thanks for the fast reply.
i/m a newbiew in the CRS amd Modsecuirty rule syntax.

Where to insert your exclusion rule?

**Comment by azurit** [comment]
Into file REQUEST-900-EXCLUSION-RULES-BEFORE-CRS.conf

**Comment by albgen** [comment]
It has triggered the following...

`ModSecurity: Warning. detected XSS using libinjection. [file "/etc/nginx/modsec/owasp-modsecurity-crs/rules/REQUEST-941-APPLICATION-ATTACK-XSS.conf"] [line "37"] [id "941100"] [rev ""] [msg "XSS Attack Detected via libinjection"] [data "Matched Data: XSS data found within ARGS:modules: [{"type":"et_pb_section","attrs":{"next_background_color":"#FFFFFF","disabled_on":"on|on|off","admin_label":"Hero","_builder_version":"4.4.1","backgro (30139 characters omitted)"] [severity "2"] [ver "OWASP_CRS/3.2.0"] [maturity "0"] [accuracy "0"] [tag "application-multi"] [tag "language-multi"] [tag "platform-multi"] [tag "attack-xss"] [tag "paranoia-level/1"] [tag "OWASP_CRS"] [tag "OWASP_CRS/WEB_ATTACK/XSS"] [tag "WASCTC/WASC-8"] [tag "WASCTC/WASC-22"] [tag "OWASP_TOP_10/A3"] [tag "OWASP_AppSensor/IE1"] [tag "CAPEC-242"] [hostname "5.249.X.X"] [uri "/wp-admin/admin-ajax.php"] [unique_id "158522966188.995404"] [ref "v4169,30487t:utf8toUnicode,t:urlDecodeUni,t:htmlEntityDecode,t:jsDecode,t:cssDecode,t:removeNulls"]
ModSecurity: Warning. Matched "Operator `Rx' with parameter `(?i:(?:<\w[\s\S]*[\s\/]|['\"](?:[\s\S]*[\s\/])?)(?:on(?:d(?:e(?:vice(?:(?:orienta|mo)tion|proximity|found|light)|livery(?:success|error)|activate)|r(?:ag(?:e(?:n(?:ter|d)|xit)|(?:gestur|leav)e|start|d (3146 characters omitted)' against variable `ARGS:modules' (Value: `[{"type":"et_pb_section","attrs":{"next_background_color":"#FFFFFF","disabled_on":"on|on|off","admin (31287 characters omitted)' ) [file "/etc/nginx/modsec/owasp-modsecurity-crs/rules/REQUEST-941-APPLICATION-ATTACK-XSS.conf"] [line "205"] [id "941160"] [rev ""] [msg "NoScript XSS InjectionChecker: HTML Injection"] [data "Matched Data: "type":"et_pb_section","attrs":{"next_background_color":"#FFFFFF","disabled_on":"on|on|off","admin_label":"Hero","_builder_version":"4.4.1","background_enable_color":"off","custom_paddin (51159 characters omitted)"] [severity "2"] [ver "OWASP_CRS/3.2.0"] [maturity "0"] [accuracy "0"] [tag "application-multi"] [tag "language-multi"] [tag "platform-multi"] [tag "attack-xss"] [tag "paranoia-level/1"] [tag "OWASP_CRS"] [tag "OWASP_CRS/WEB_ATTACK/XSS"] [tag "WASCTC/WASC-8"] [tag "WASCTC/WASC-22"] [tag "OWASP_TOP_10/A3"] [tag "OWASP_AppSensor/IE1"] [tag "CAPEC-242"] [hostname "5.249.X.X"] [uri "/wp-admin/admin-ajax.php"] [unique_id "158522966188.995404"] [ref "o2,21028v4169,30487t:utf8toUnicode,t:urlDecodeUni,t:htmlEntityDecode,t:jsDecode,t:cssDecode,t:removeNulls"]
ModSecurity: Access denied with code 403 (phase 2). Matched "Operator `Ge' with parameter `5' against variable `TX:ANOMALY_SCORE' (Value: `10' ) [file "/etc/nginx/modsec/owasp-modsecurity-crs/rules/REQUEST-949-BLOCKING-EVALUATION.conf"] [line "79"] [id "949110"] [rev ""] [msg "Inbound Anomaly Score Exceeded (Total Score: 10)"] [data ""] [severity "2"] [ver ""] [maturity "0"] [accuracy "0"] [tag "application-multi"] [tag "language-multi"] [tag "platform-multi"] [tag "attack-generic"] [hostname "5.249.X.X"] [uri "/wp-admin/admin-ajax.php"] [unique_id "158522966188.995404"] [ref ""]`

**Comment by azurit** [comment]
Try adding also this rule:

```
SecRule REQUEST_FILENAME "@endsWith /wp-admin/admin-ajax.php" \
    "id:9990127,\
    phase:2,\
    pass,\
    t:none,\
    nolog,\
    chain"
    SecRule ARGS_POST:action "@rx ^(?:et_fb_get_shortcode_from_fb_object|et_fb_ajax_save)$" \
        "t:none,\
        chain"
        SecRule &ARGS_POST:action "@eq 1" \
            "t:none,\
            ctl:ruleRemoveTargetById=932150;ARGS:modules,\
            ctl:ruleRemoveTargetById=942230;ARGS:modules,\
            ctl:ruleRemoveTargetById=932100;ARGS:modules,\
            ctl:ruleRemoveTargetById=942170;ARGS:modules,\
            ctl:ruleRemoveTargetById=941130;ARGS:modules,\
            ctl:ruleRemoveTargetById=941170;ARGS:modules,\
            ctl:ruleRemoveTargetById=932130;ARGS:modules,\
            ctl:ruleRemoveTargetById=941120;ARGS:modules,\
            ctl:ruleRemoveTargetById=934100;ARGS:modules,\
            ctl:ruleRemoveTargetById=941100;ARGS:modules,\
            ctl:ruleRemoveTargetById=941110;ARGS:modules,\
            ctl:ruleRemoveTargetById=941160;ARGS:modules,\
            ctl:ruleRemoveTargetById=941180;ARGS:modules"

```

**Comment by albgen** [comment]
Now i can save the pages.
Is there other workaround rather than excluding the rules?

thanks

**Comment by azurit** [comment]
No.

**Comment by azurit** [comment]
With this, you are NOT excluding the whole rules on the whole server, just on very little part of the requests and only for specified POST parameters. It is safe to do this and has no effect on other types of requests.

**Comment by albgen** [comment]
> Try adding also this rule:
> 
> ```
> SecRule REQUEST_FILENAME "@endsWith /wp-admin/admin-ajax.php" \
>     "id:9990127,\
>     phase:2,\
>     pass,\
>     t:none,\
>     nolog,\
>     chain"
>     SecRule ARGS_POST:action "@rx ^(?:et_fb_get_shortcode_from_fb_object|et_fb_ajax_save)$" \
>         "t:none,\
>         chain"
>         SecRule &ARGS_POST:action "@eq 1" \
>             "t:none,\
>             ctl:ruleRemoveTargetById=932150;ARGS:modules,\
>             ctl:ruleRemoveTargetById=942230;ARGS:modules,\
>             ctl:ruleRemoveTargetById=932100;ARGS:modules,\
>             ctl:ruleRemoveTargetById=942170;ARGS:modules,\
>             ctl:ruleRemoveTargetById=941130;ARGS:modules,\
>             ctl:ruleRemoveTargetById=941170;ARGS:modules,\
>             ctl:ruleRemoveTargetById=932130;ARGS:modules,\
>             ctl:ruleRemoveTargetById=941120;ARGS:modules,\
>             ctl:ruleRemoveTargetById=934100;ARGS:modules,\
>             ctl:ruleRemoveTargetById=941100;ARGS:modules,\
>             ctl:ruleRemoveTargetById=941110;ARGS:modules,\
>             ctl:ruleRemoveTargetById=941160;ARGS:modules,\
>             ctl:ruleRemoveTargetById=941180;ARGS:modules"
> ```



> With this, you are NOT excluding the whole rules on the whole server, just on very little part of the requests and only for specified POST parameters. It is safe to do this and has no effect on other types of requests.

i think it has disabled other rules as well.
for example even a simple rule like 

>  Basic test rule
> 		SecRule ARGS:testparam "@contains test" "id:1234,deny,status:403"

is not working anymore: curl https://domain/?testparam=test		
If i remove your rule it works again.

Even with a simple XSS request it goes through
curl https://domain/?param="><script>alert(1);</script>"

**Comment by azurit** [comment]
Then you did something wrong.

**Comment by albgen** [comment]
just added the rule(the second one you posted) in REQUEST-900-EXCLUSION-RULES-BEFORE-CRS.conf The file contains only that rule.
Restart nginx and thats it.

No idea why seems the whole WAF is disabled...will try to understand better whats happening :(

**Comment by azurit** [comment]
Maybe you copied something wrong (some syntax errors in config file) and your nginx disabled modsecurity? Not using nginx so i don't know how it works.

Also, my rules are for Apache and i don't know if the syntax is the same.

**Comment by albgen** [comment]
Syntax is ok. nginx does not even start if there is a syntax problem. Here is a sample if i put gibberish on a config file in REQUEST-900-EXCLUSION-RULES-BEFORE-CRS.conf

So the syntax of your rule is correct. Now, why it seems has disabled is another story. Maybe i should learn how to debug modsec :)


[root@web ~]# nginx -t
nginx: [emerg] "modsecurity_rules_file" directive Rules error. File: /etc/nginx/modsec/owasp-modsecurity-crs/rules/REQUEST-900-EXCLUSION-RULES-BEFORE-CRS.conf. Line: 9. Column: 11. **Invalid input:  aaaaaaaaaa** in /etc/nginx/nginx.conf:35
nginx: configuration file /etc/nginx/nginx.conf test failed

**Comment by azurit** [comment]
So i cannot help you more. That rule is a part of our own rules which we are using on all of our webhosting servers (we are one of the lead hosting providers in Slovakia).

**Comment by albgen** [comment]
ok i see. thanks man.
Will let you know my findings
